## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [2]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [3]:
pwd

'C:\\Users\\User\\Downloads'

In [43]:
# Write your code for importing the csv file here
movies =pd.read_csv("Movies.csv")


In [44]:
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [6]:
# Write your code for inspection here


#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [7]:
movies.shape

(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [8]:
a=movies.isnull().sum()
b=a[a[:]!=0]
b

color                      2
num_critic_for_reviews     1
duration                   1
actor_3_facebook_likes     6
actor_2_name               1
actor_3_name               6
facenumber_in_poster       6
plot_keywords             30
language                   3
content_rating            48
actor_2_facebook_likes     1
aspect_ratio              72
dtype: int64

In [9]:
b.max()

72

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [10]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
movies.drop(["color","director_facebook_likes","actor_1_facebook_likes","actor_2_facebook_likes","actor_3_facebook_likes","actor_2_name","cast_total_facebook_likes","actor_3_name","duration","facenumber_in_poster","content_rating","country","movie_imdb_link","aspect_ratio","plot_keywords"], axis=1)


director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [11]:
# Write you code here
b

color                      2
num_critic_for_reviews     1
duration                   1
actor_3_facebook_likes     6
actor_2_name               1
actor_3_name               6
facenumber_in_poster       6
plot_keywords             30
language                   3
content_rating            48
actor_2_facebook_likes     1
aspect_ratio              72
dtype: int64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [12]:
# Write your code for filling the NaN values in the 'language' column here
movies.loc[pd.isnull(movies['language']), ['language']] = 'English'
(movies.language == 'English').sum()

3674

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [13]:
# Write your code for unit conversion here
movies.head()
movies.loc[:, ["gross","budget"]]=movies.loc[:, ["gross","budget"]]/10000000
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes      gross                               genres  \
0                     1000.0  76.050585      Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  30.940415             Action|Adventure|Fantasy   
2                    11000.0  20.007417            Action|Adventure|Thriller   
3                    27000.0  44.813064                      Action|Thriller   
4                      640.0   7.305868              Action|Adventure|Sci-Fi   
...                      ...        ...                                  ...   
3848                   291.0   0.042476                Drama|Sci-Fi|Thriller   
3849                     0.0   0.007007                             Thriller   
3850                   121.0   0.204092  Action|Crime|Drama|Romance|Thriller   
3851                   296.0   0.000458                         Comedy|Drama   
3852                    86.0   0.008522                          Documentary   

      ... num_user_for_reviews language      country  content_rating  \
0     ...               3054.0  English          USA           PG-13   
1     ...               1238.0  English          USA           PG-13   
2     ...                994.0  English           UK           PG-13   
3     ...               2701.0  English          USA           PG-13   
4     ...                738.0  English          USA           PG-13   
...   ...                  ...      ...          ...             ...   
3848  ...                371.0  English          USA           PG-13   
3849  ...                 35.0  English  Philippines       Not Rated   
3850  ...                130.0  Spanish          USA               R   
3851  ...                 14.0  English          USA       Not Rated   
3852  ...                 84.0  English          USA              PG   

        budget  title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0     23.70000      2009.0                  936.0        7.9          1.78   
1     30.00000      2007.0                 5000.0        7.1          2.35   
2     24.50000      2015.0                  393.0        6.8          2.35   
3     25.00000      2012.0                23000.0        8.5          2.35   
4     26.37000      2012.0                  632.

In [14]:
movies.loc[:, ["budget"]]=movies.loc[:, ["budget"]]/10000000
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes      gross                               genres  \
0                     1000.0  76.050585      Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  30.940415             Action|Adventure|Fantasy   
2                    11000.0  20.007417            Action|Adventure|Thriller   
3                    27000.0  44.813064                      Action|Thriller   
4                      640.0   7.305868              Action|Adventure|Sci-Fi   
...                      ...        ...                                  ...   
3848                   291.0   0.042476                Drama|Sci-Fi|Thriller   
3849                     0.0   0.007007                             Thriller   
3850                   121.0   0.204092  Action|Crime|Drama|Romance|Thriller   
3851                   296.0   0.000458                         Comedy|Drama   
3852                    86.0   0.008522                          Documentary   

      ... num_user_for_reviews language      country  content_rating  \
0     ...               3054.0  English          USA           PG-13   
1     ...               1238.0  English          USA           PG-13   
2     ...                994.0  English           UK           PG-13   
3     ...               2701.0  English          USA           PG-13   
4     ...                738.0  English          USA           PG-13   
...   ...                  ...      ...          ...             ...   
3848  ...                371.0  English          USA           PG-13   
3849  ...                 35.0  English  Philippines       Not Rated   
3850  ...                130.0  Spanish          USA               R   
3851  ...                 14.0  English          USA       Not Rated   
3852  ...                 84.0  English          USA              PG   

            budget  title_year actor_2_facebook_likes imdb_score  \
0     2.370000e-06      2009.0                  936.0        7.9   
1     3.000000e-06      2007.0                 5000.0        7.1   
2     2.450000e-06      2015.0                  393.0        6.8   
3     2.500000e-06      2012.0                23000.0        8.5   
4     2.637000e-06      2012.0                  632.0        6.6   
...            ...         ...

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [15]:
# Write your code for creating the profit column here
movies['profit']=movies["gross"]-movies['budget']
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes      gross                               genres  \
0                     1000.0  76.050585      Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  30.940415             Action|Adventure|Fantasy   
2                    11000.0  20.007417            Action|Adventure|Thriller   
3                    27000.0  44.813064                      Action|Thriller   
4                      640.0   7.305868              Action|Adventure|Sci-Fi   
...                      ...        ...                                  ...   
3848                   291.0   0.042476                Drama|Sci-Fi|Thriller   
3849                     0.0   0.007007                             Thriller   
3850                   121.0   0.204092  Action|Crime|Drama|Romance|Thriller   
3851                   296.0   0.000458                         Comedy|Drama   
3852                    86.0   0.008522                          Documentary   

      ... language      country  content_rating        budget title_year  \
0     ...  English          USA           PG-13  2.370000e-06     2009.0   
1     ...  English          USA           PG-13  3.000000e-06     2007.0   
2     ...  English           UK           PG-13  2.450000e-06     2015.0   
3     ...  English          USA           PG-13  2.500000e-06     2012.0   
4     ...  English          USA           PG-13  2.637000e-06     2012.0   
...   ...      ...          ...             ...           ...        ...   
3848  ...  English          USA           PG-13  7.000000e-11     2004.0   
3849  ...  English  Philippines       Not Rated  7.000000e-11     2005.0   
3850  ...  Spanish          USA               R  7.000000e-11     1992.0   
3851  ...  English          USA       Not Rated  9.000000e-11     2011.0   
3852  ...  English          USA              PG  1.100000e-11     2004.0   

      actor_2_facebook_likes imdb_score aspect_ratio  movie_facebook_likes  \
0                      936.0        7.9         1.78                 33000   
1                     5000.0        7.1         2.35                     0   
2                      393.0        6.8         2.35                 85000   
3                    23000.0        8.5         2.35                164000   


In [16]:
# Write your code for sorting the dataframe here
movies.sort_values(by='profit', ascending=False)

color          director_name  num_critic_for_reviews  duration  \
0     Color          James Cameron                   723.0     178.0   
25    Color          James Cameron                   315.0     194.0   
28    Color        Colin Trevorrow                   644.0     124.0   
16    Color            Joss Whedon                   703.0     173.0   
64    Color      Christopher Nolan                   645.0     152.0   
...     ...                    ...                     ...       ...   
3742  Color        Alex Craig Mann                    29.0      87.0   
3805  Color            Ricki Stern                    11.0     106.0   
3670  Color  Brian Trenchard-Smith                     8.0      88.0   
3671  Color           Frank Whaley                     9.0      96.0   
2925  Color    Ekachai Uekrongtham                    66.0      96.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
25                        0.0                   794.0       Kate Winslet   
28                      365.0                  1000.0         Judy Greer   
16                        0.0                 19000.0  Robert Downey Jr.   
64                    22000.0                 11000.0       Heath Ledger   
...                       ...                     ...                ...   
3742                     38.0                   445.0   Christa B. Allen   
3805                     15.0                     0.0   Evelyn Jefferson   
3670                     53.0                   176.0   Mariel Hemingway   
3671                    436.0                     4.0       Frank Whaley   
2925                      3.0                   305.0         Mike Dopud   

      actor_1_facebook_likes      gross                            genres  \
0                     1000.0  76.050585   Action|Adventure|Fantasy|Sci-Fi   
25                   29000.0  65.867230                     Drama|Romance   
28                    3000.0  65.217727  Action|Adventure|Sci-Fi|Thriller   
16                   26000.0  62.327955           Action|Adventure|Sci-Fi   
64                   23000.0  53.331606       Action|Crime|Drama|Thriller   
...                      ...        ...                               ...   
3742                   552.0   0.000133                     Comedy|Horror   
3805                     2.0   0.000111                 Crime|Documentary   
3670                   563.0   0.000072                      Action|Drama   
3671                   474.0   0.000070                      Comedy|Drama   
2925                  2000.0   0.000016             Action|Crime|Thriller   

      ... language   country  content_rating        budget title_year  \
0     ...  English       USA           PG-13  2.370000e-06     2009.0   
25    ...  English       USA           PG-13  2.000000e-06     1997.0   
28    ...  English       USA           PG-13  1.500000e-06     2015.0   
16    ...  English       USA           PG-13  2.200000e-06     2012.0   
64    ...  English       USA           PG-13  1.850000e-06     2008.0   
...   ...      ...       ...             ...           ...        ...   
3742  ...  English       USA       Not Rated  5.000000e-09     2012.0   
3805  ...  English       USA           PG-13  2.000000e-09     2006.0   
3670  ...  English   Germany               R  1.000000e-08     2006.0   
3671  ...  English       USA               R  1.500000e-08     2001.0   
2925  ...  English  Thailand               R  9.000000e-08     2014.0   

      actor_2_facebook_likes imdb_score aspect_ratio  movie_facebook_likes  \
0                      936.0        7.9         1.78                 33000   
25                   14000.0        7.7         2.35                 26000   
28                    2000.0        7.0         2.00                150000   
16                   21000.0        8.1         1.85                123000   
64                   13000.0        

In [17]:
top10 =movies.sort_values(by='profit', ascending=False).head(5)
top10["movie_title"] # Write your code to get the top 10 profiting movies here

0              Avatar 
25            Titanic 
28     Jurassic World 
16       The Avengers 
64    The Dark Knight 
Name: movie_title, dtype: object

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [18]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250=movies[movies['num_voted_users']>25000].sort_values(by='imdb_score', ascending=False).head(250)

IMDb_Top_250['Rank'] = range(1,251)
IMDb_Top_250

color         director_name  num_critic_for_reviews  duration  \
1795  Color        Frank Darabont                   199.0     142.0   
3016  Color  Francis Ford Coppola                   208.0     175.0   
64    Color     Christopher Nolan                   645.0     152.0   
2543  Color  Francis Ford Coppola                   149.0     220.0   
325   Color         Peter Jackson                   328.0     192.0   
...     ...                   ...                     ...       ...   
67    Color           Jon Favreau                   486.0     126.0   
73    Color            Doug Liman                   585.0     113.0   
3052  Color           Yash Chopra                    29.0     192.0   
95    Color         Peter Jackson                   645.0     182.0   
3170  Color  Christophe Barratier                   112.0      97.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
1795                      0.0                   461.0     Jeffrey DeMunn   
3016                      0.0                  3000.0      Marlon Brando   
64                    22000.0                 11000.0       Heath Ledger   
2543                      0.0                  3000.0          Al Pacino   
325                       0.0                   416.0         Billy Boyd   
...                       ...                     ...                ...   
67                     4000.0                  4000.0       Jeff Bridges   
73                      218.0                   509.0        Lara Pulver   
3052                    147.0                   397.0       Preity Zinta   
95                        0.0                   773.0         Adam Brown   
3170                     25.0                    57.0  François Berléand   

      actor_1_facebook_likes      gross                          genres  ...  \
1795                 11000.0   2.834147                     Crime|Drama  ...   
3016                 14000.0  13.482195                     Crime|Drama  ...   
64                   23000.0  53.331606     Action|Crime|Drama|Thriller  ...   
2543                 22000.0   5.730000                     Crime|Drama  ...   
325                   5000.0  37.701925  Action|Adventure|Drama|Fantasy  ...   
...                      ...        ...                             ...  ...   
67                   21000.0  31.829818         Action|Adventure|Sci-Fi  ...   
73                   10000.0  10.018950         Action|Adventure|Sci-Fi  ...   
3052                  8000.0   0.292174           Drama|Musical|Romance  ...   
95                    5000.0  30.300123               Adventure|Fantasy  ...   
3170                   517.0   0.362976                     Drama|Music  ...   

     country content_rating        budget  title_year actor_2_facebook_likes  \
1795     USA              R  2.500000e-07      1994.0                  745.0   
3016     USA              R  6.000000e-08      1972.0                10000.0   
64       USA          PG-13  1.850000e-06      2008.0                13000.0   
2543     USA              R  1.300000e-07      1974.0                14000.0   
325      USA          PG-13  9.400000e-07      2003.0                  857.0   
...      ...            ...           ...         ...                    ...   
67       USA          PG-13  1.400000e-06      2008.0                12000.0   
73       USA          PG-13  1.780000e-06      2014.0                  854.0   
3052   India            NaN  7.000000e-08      2004.0                  860.0   
95       USA          PG-13  1.800000e-06      2012.0                  972.0   
3170  France          PG-13  5.500000e-08      2004.0                   86.0   

      imdb_score aspect_ratio movie_facebook_likes     profit Rank  
1795         9.3         1.85               108000   2.834147    1  
3016         9.2         1.85                43000  13.482195    2  
64           9.0         2.35                37000  53.331604    3  
2543         9.0         1.85             

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [19]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=7.5) & (IMDb_Top_250['imdb_score']<=8)]

color         director_name  num_critic_for_reviews  duration  \
2454  Color       Martin McDonagh                   300.0     107.0   
2444  Color         Terry Gilliam                   230.0     142.0   
2360  Color            Tom Hooper                   479.0     118.0   
2156  Color         Kevin Costner                    92.0     236.0   
2706  Color            David Lean                    89.0     200.0   
...     ...                   ...                     ...       ...   
67    Color           Jon Favreau                   486.0     126.0   
73    Color            Doug Liman                   585.0     113.0   
3052  Color           Yash Chopra                    29.0     192.0   
95    Color         Peter Jackson                   645.0     182.0   
3170  Color  Christophe Barratier                   112.0      97.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
2454                    454.0                    39.0       Anna Madeley   
2444                      0.0                  1000.0        Bob Hoskins   
2360                      0.0                   520.0      Jennifer Ehle   
2156                      0.0                   232.0     Michael Spears   
2706                    767.0                   382.0       Klaus Kinski   
...                       ...                     ...                ...   
67                     4000.0                  4000.0       Jeff Bridges   
73                      218.0                   509.0        Lara Pulver   
3052                    147.0                   397.0       Preity Zinta   
95                        0.0                   773.0         Adam Brown   
3170                     25.0                    57.0  François Berléand   

      actor_1_facebook_likes      gross                           genres  ...  \
2454                    65.0   0.775713               Comedy|Crime|Drama  ...   
2444                 22000.0   0.992900                     Drama|Sci-Fi  ...   
2360                 14000.0  13.879534  Biography|Drama|History|Romance  ...   
2156                   933.0  18.420885          Adventure|Drama|Western  ...   
2706                   597.0  11.172200                Drama|Romance|War  ...   
...                      ...        ...                              ...  ...   
67                   21000.0  31.829818          Action|Adventure|Sci-Fi  ...   
73                   10000.0  10.018950          Action|Adventure|Sci-Fi  ...   
3052                  8000.0   0.292174            Drama|Musical|Romance  ...   
95                    5000.0  30.300123                Adventure|Fantasy  ...   
3170                   517.0   0.362976                      Drama|Music  ...   

     country content_rating        budget  title_year actor_2_facebook_likes  \
2454      UK              R  1.500000e-07      2008.0                   49.0   
2444      UK              R  1.500000e-07      1985.0                 5000.0   
2360      UK              R  1.500000e-07      2010.0                 1000.0   
2156     USA          PG-13  2.200000e-07      1990.0                  839.0   
2706     USA          PG-13  1.100000e-07      1965.0                  396.0   
...      ...            ...           ...         ...                    ...   
67       USA          PG-13  1.400000e-06      2008.0                12000.0   
73       USA          PG-13  1.780000e-06      2014.0                  854.0   
3052   India            NaN  7.000000e-08      2004.0                  860.0   
95       USA          PG-13  1.800000e-06      2012.0                  972.0   
3170  France          PG-13  5.500000e-08      2004.0                   86.0   

      imdb_score aspect_ratio movie_facebook_likes     profit Rank  
2454         8.0         2.35                32000   0.775713  153  
2444         8.0         1.85                16000   0.992900  154  
2360         8.0         1.85                64000  13.879534  155  
2156         8.0         2.35 

In [20]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=8) & (IMDb_Top_250['imdb_score']<=8.5)]

color         director_name  num_critic_for_reviews  \
1467             Color  Francis Ford Coppola                   261.0   
3198   Black and White     Christopher Nolan                   274.0   
1360   Black and White        Roman Polanski                   193.0   
1159             Color     Christopher Nolan                   341.0   
3377             Color       Damien Chazelle                   535.0   
...                ...                   ...                     ...   
3445   Black and White            Mel Brooks                   129.0   
3009             Color     Vincent Paronnaud                   242.0   
1253             Color           Joss Whedon                   276.0   
3307             Color     Richard Linklater                   548.0   
3511             Color            Ari Folman                   231.0   

      duration  director_facebook_likes  actor_3_facebook_likes  \
1467     289.0                      0.0                  3000.0   
3198     113.0                  22000.0                   379.0   
1360     150.0                   2000.0                    95.0   
1159     130.0                  22000.0                 19000.0   
3377     107.0                    141.0                   535.0   
...        ...                      ...                     ...   
3445     106.0                      0.0                   481.0   
3009      89.0                     10.0                   106.0   
1253     119.0                      0.0                   279.0   
3307     165.0                      0.0                   127.0   
3511      90.0                     56.0                     0.0   

           actor_2_name  actor_1_facebook_likes     gross  \
1467      Marlon Brando                 11000.0  7.880000   
3198      Thomas Lennon                   716.0  2.553088   
1360       Frank Finlay                   396.0  3.251932   
1159       Hugh Jackman                 23000.0  5.308274   
3377    Melissa Benoist                 24000.0  1.309200   
...                 ...                     ...       ...   
3445        Peter Boyle                  1000.0  8.630000   
3009      Gena Rowlands                   963.0  0.444340   
1253         Sean Maher                  2000.0  2.533593   
3307  Lorelei Linklater                   230.0  2.535920   
3511        Ronny Dayag                    56.0  0.228328   

                                                 genres  ... country  \
1467                                          Drama|War  ...     USA   
3198                                   Mystery|Thriller  ...     USA   
1360                                Biography|Drama|War  ...  France   
1159                      Drama|Mystery|Sci-Fi|Thriller  ...     USA   
3377                                        Drama|Music  ...     USA   
...                                                 ...  ...     ...   
3445                                             Comedy  ...     USA   
3009                      Animation|Biography|Drama|War  ...  France   
1253                   Action|Adventure|Sci-Fi|Thriller  ...     USA   
3307                                              Drama  ...     USA   
3511  Animation|Biography|Documentary|Drama|History|War  ...  Israel   

     content_rating        budget  title_year actor_2_facebook_likes  \
1467              R  3.150000e-07      1979.0                10000.0   
3198              R  9.000000e-08      2000.0                  651.0   
1360              R  3.500000e-07      2002.0                  338.0   
1159          PG-13  4.000000e-07      2006.0                20000.0   
3377              R  3.300000e-08      2014.0                  970.0   
...             ...           ...         ...                    ...   
3445             PG  2.800000e-08      1974.0                  595.0   
3009          PG-13  7.300000e-08      2007.0                  545.0   
1253          PG-13  4.000000e-07      2005.0                  505.0   
3307              R  4.000000

In [21]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=8.5) & (IMDb_Top_250['imdb_score']<=9)].shape

(44, 30)

In [22]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=9) & (IMDb_Top_250['imdb_score']<=9.5)]

color         director_name  num_critic_for_reviews  duration  \
1795  Color        Frank Darabont                   199.0     142.0   
3016  Color  Francis Ford Coppola                   208.0     175.0   
64    Color     Christopher Nolan                   645.0     152.0   
2543  Color  Francis Ford Coppola                   149.0     220.0   

      director_facebook_likes  actor_3_facebook_likes    actor_2_name  \
1795                      0.0                   461.0  Jeffrey DeMunn   
3016                      0.0                  3000.0   Marlon Brando   
64                    22000.0                 11000.0    Heath Ledger   
2543                      0.0                  3000.0       Al Pacino   

      actor_1_facebook_likes      gross                       genres  ...  \
1795                 11000.0   2.834147                  Crime|Drama  ...   
3016                 14000.0  13.482195                  Crime|Drama  ...   
64                   23000.0  53.331606  Action|Crime|Drama|Thriller  ...   
2543                 22000.0   5.730000                  Crime|Drama  ...   

     country content_rating        budget  title_year actor_2_facebook_likes  \
1795     USA              R  2.500000e-07      1994.0                  745.0   
3016     USA              R  6.000000e-08      1972.0                10000.0   
64       USA          PG-13  1.850000e-06      2008.0                13000.0   
2543     USA              R  1.300000e-07      1974.0                14000.0   

      imdb_score aspect_ratio movie_facebook_likes     profit Rank  
1795         9.3         1.85               108000   2.834147    1  
3016         9.2         1.85                43000  13.482195    2  
64           9.0         2.35                37000  53.331604    3  
2543         9.0         1.85                14000   5.730000    4  

[4 rows x 30 columns]

In [23]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=9.5) & (IMDb_Top_250['imdb_score']<=10)]

Empty DataFrame
Columns: [color, director_name, num_critic_for_reviews, duration, director_facebook_likes, actor_3_facebook_likes, actor_2_name, actor_1_facebook_likes, gross, genres, actor_1_name, movie_title, num_voted_users, cast_total_facebook_likes, actor_3_name, facenumber_in_poster, plot_keywords, movie_imdb_link, num_user_for_reviews, language, country, content_rating, budget, title_year, actor_2_facebook_likes, imdb_score, aspect_ratio, movie_facebook_likes, profit, Rank]
Index: []

[0 rows x 30 columns]

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [52]:
# Write your code for creating three new dataframes here
Meryl_Streep=movies[movies['actor_1_name']== 'Meryl Streep'] 
Meryl_Streep.head()

color   director_name  num_critic_for_reviews  duration  \
392   Color    Nancy Meyers                   187.0     120.0   
1038  Color   Curtis Hanson                    42.0     111.0   
1132  Color     Nora Ephron                   252.0     123.0   
1322  Color   David Frankel                   208.0     109.0   
1390  Color  Robert Redford                   227.0      92.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
392                     278.0                   963.0     Hunter Parrish   
1038                    161.0                   132.0    Glenn Morshower   
1132                      0.0                   923.0  Mary Lynn Rajskub   
1322                     64.0                   505.0      Anne Hathaway   
1390                      0.0                 10000.0         Tom Cruise   

      actor_1_facebook_likes        gross                           genres  \
392                  11000.0  112703470.0             Comedy|Drama|Romance   
1038                 11000.0   46815748.0  Action|Adventure|Crime|Thriller   
1132                 11000.0   94125426.0          Biography|Drama|Romance   
1322                 11000.0  124732962.0             Comedy|Drama|Romance   
1390                 11000.0   14998070.0               Drama|Thriller|War   

      ... num_user_for_reviews language  country  content_rating      budget  \
392   ...                214.0  English      USA               R  85000000.0   
1038  ...                 69.0  English      USA           PG-13  45000000.0   
1132  ...                277.0  English      USA           PG-13  40000000.0   
1322  ...                631.0  English      USA           PG-13  35000000.0   
1390  ...                298.0  English      USA               R  35000000.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
392       2009.0                 2000.0        6.6          1.85   
1038      1994.0                  894.0        6.3          2.35   
1132      2009.0                  935.0        7.0          1.85   
1322      2006.0                11000.0        6.8          2.35   
1390      2007.0                10000.0        6.2          2.35   

     movie_facebook_likes  
392                     0  
1038                    0  
1132                13000  
1322                    0  
1390                    0  

[5 rows x 28 columns]

In [48]:
Leo_Caprio =movies[movies['actor_1_name']== "Leonardo DiCaprio"]
Leo_Caprio.head()

color          director_name  num_critic_for_reviews  \
25              Color          James Cameron                   315.0   
49              Color           Baz Luhrmann                   490.0   
94              Color      Christopher Nolan                   642.0   
173             Color  Alejandro G. Iñárritu                   556.0   
246   Black and White        Martin Scorsese                   267.0   

     duration  director_facebook_likes  actor_3_facebook_likes  \
25      194.0                      0.0                   794.0   
49      143.0                   1000.0                    77.0   
94      148.0                  22000.0                 23000.0   
173     156.0                      0.0                   733.0   
246     170.0                  17000.0                   827.0   

          actor_2_name  actor_1_facebook_likes        gross  \
25        Kate Winslet                 29000.0  658672302.0   
49   Elizabeth Debicki                 29000.0  144812796.0   
94           Tom Hardy                 29000.0  292568851.0   
173          Tom Hardy                 29000.0  183635922.0   
246         Adam Scott                 29000.0  102608827.0   

                               genres  ... num_user_for_reviews language  \
25                      Drama|Romance  ...               2528.0  English   
49                      Drama|Romance  ...                753.0  English   
94   Action|Adventure|Sci-Fi|Thriller  ...               2803.0  English   
173  Adventure|Drama|Thriller|Western  ...               1188.0  English   
246                   Biography|Drama  ...                799.0  English   

       country  content_rating       budget  title_year  \
25         USA           PG-13  200000000.0      1997.0   
49   Australia           PG-13  105000000.0      2013.0   
94         USA           PG-13  160000000.0      2010.0   
173        USA               R  135000000.0      2015.0   
246        USA           PG-13  110000000.0      2004.0   

    actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
25                 14000.0        7.7          2.35                26000  
49                   509.0        7.3          2.35               115000  
94                 27000.0        8.8          2.35               175000  
173                27000.0        8.1          2.35               190000  
246                 3000.0        7.5          2.35                    0  

[5 rows x 28 columns]

In [50]:
Brad_Pitt =movies[movies['actor_1_name']== "Brad Pitt"] 
Brad_Pitt.head()

color      director_name  num_critic_for_reviews  duration  \
97   Color      David Fincher                   362.0     166.0   
142  Color  Wolfgang Petersen                   220.0     196.0   
243  Color  Steven Soderbergh                   198.0     125.0   
244  Color         Doug Liman                   233.0     126.0   
367  Color         Tony Scott                   142.0     114.0   

     director_facebook_likes  actor_3_facebook_likes         actor_2_name  \
97                   21000.0                   919.0        Jason Flemyng   
142                    249.0                   844.0        Orlando Bloom   
243                      0.0                   350.0        Julia Roberts   
244                    218.0                   322.0  Angelina Jolie Pitt   
367                  12000.0                   307.0      Stephen Dillane   

     actor_1_facebook_likes        gross  \
97                  11000.0  127490802.0   
142                 11000.0  133228348.0   
243                 11000.0  125531634.0   
244                 11000.0  186336103.0   
367                 11000.0      26871.0   

                                   genres  ... num_user_for_reviews language  \
97                  Drama|Fantasy|Romance  ...                822.0  English   
142                             Adventure  ...               1694.0  English   
243                        Crime|Thriller  ...                627.0  English   
244  Action|Comedy|Crime|Romance|Thriller  ...                798.0  English   
367                 Action|Crime|Thriller  ...                361.0  English   

     country  content_rating       budget  title_year actor_2_facebook_likes  \
97       USA           PG-13  150000000.0      2008.0                 1000.0   
142      USA               R  175000000.0      2004.0                 5000.0   
243      USA           PG-13  110000000.0      2004.0                 8000.0   
244      USA           PG-13  120000000.0      2005.0                11000.0   
367  Germany               R   92000000.0      2001.0                  577.0   

    imdb_score  aspect_ratio movie_facebook_likes  
97         7.8          2.35                23000  
142        7.2          2.35                    0  
243        6.4          2.35                    0  
244        6.5          2.35                    0  
367        7.0          2.35                    0  

[5 rows x 28 columns]

In [53]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt], axis=0)
Combined.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [54]:
# Write your code for grouping the combined dataframe here
Combined.groupby(['actor_1_name']).agg({'num_critic_for_reviews': 'mean', 'num_user_for_reviews': 'mean'})

num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

In [30]:
# Write the code for finding the mean of critic reviews and audience reviews here

Help on built-in function dir in module builtins:

dir(...)
    dir([object]) -> list of strings
    
    If called without an argument, return the names in the current scope.
    Else, return an alphabetized list of names comprising (some of) the attributes
    of the given object, and of attributes reachable from it.
    If the object supplies a method named __dir__, it will be used; otherwise
    the default dir() logic is used and returns:
      for a module object: the module's attributes.
      for a class object:  its attributes, and recursively the attributes
        of its bases.
      for any other object: its attributes, its class's attributes, and
        recursively the attributes of its class's base classes.



#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>